api/draft_registrations/serializers.py

```diff
@@ -16,17 +16,19 @@
     NodeContributorsSerializer,
     NodeContributorsCreateSerializer,
     NodeContributorDetailSerializer,
+    RegistrationSchemaRelationshipField,
 )
+
 from api.taxonomies.serializers import TaxonomizableSerializerMixin
 from osf.exceptions import DraftRegistrationStateError
+from osf.models import Node
 from website import settings
 
 
 class NodeRelationshipField(RelationshipField):
 
     def to_internal_value(self, node_id):
-        node = self.context['view'].get_node(node_id=node_id) if node_id else None
-        return {'branched_from': node}
+        return {'branched_from': Node.load(node_id)}
```

Comment:
You mentioned a DRF issue with `self.context['view']` but I didn't get the details. We have tons of this `self.context['view']` called in our code base, why only this one breaks? @Johnetordoff

Reply:
Hmm this was apparently the only test fail.
```
=================================== FAILURES ===================================
__________ TestDraftRegistrationUpdateWithNode.test_cannot_edit_node ___________
api_tests/draft_registrations/views/test_draft_registration_detail.py:425: in test_cannot_edit_node
    auth=user.auth, expect_errors=True)
tests/json_api_test_app.py:125: in wrapper
    return JSONAPIWrapper.make_wrapper(self, url, method, content_type, params, **kw)
tests/json_api_test_app.py:33: in make_wrapper
    wrapper = self._gen_request(method, url, **kw)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/webtest_plus/app.py:177: in _gen_request
    **kwargs)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/webtest/app.py:750: in _gen_request
    expect_errors=expect_errors)
tests/json_api_test_app.py:89: in do_request
    expect_errors)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/webtest/app.py:620: in do_request
    res = req.get_response(app, catch_exc_info=True)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/webob/request.py:1310: in send
    application, catch_exc_info=True)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/webob/request.py:1278: in call_application
    app_iter = application(self.environ, start_response)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/webtest/lint.py:196: in lint_app
    iterator = application(environ, start_response_wrapper)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/contrib/staticfiles/handlers.py:76: in __call__
    return self.application(environ, start_response)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/wsgi.py:133: in __call__
    response = self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/base.py:130: in get_response
    response = self._middleware_chain(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/utils/deprecation.py:117: in __call__
    response = response or self.get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:49: in inner
    response = response_for_exception(request, exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:114: in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/exception.py:47: in inner
    response = get_response(request)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/core/handlers/base.py:181: in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/contextlib.py:52: in inner
    return func(*args, **kwds)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/views/decorators/csrf.py:54: in wrapped_view
    return view_func(*args, **kwargs)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/django/views/generic/base.py:70: in view
    return self.dispatch(request, *args, **kwargs)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/views.py:509: in dispatch
    response = self.handle_exception(exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/views.py:469: in handle_exception
    self.raise_uncaught_exception(exc)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/views.py:480: in raise_uncaught_exception
    raise exc
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/views.py:506: in dispatch
    response = handler(request, *args, **kwargs)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/generics.py:285: in put
    return self.update(request, *args, **kwargs)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/mixins.py:67: in update
    serializer.is_valid(raise_exception=True)
api/base/serializers.py:1615: in is_valid
    ret = super(JSONAPISerializer, self).is_valid(**kwargs)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/serializers.py:227: in is_valid
    self._validated_data = self.run_validation(self.initial_data)
api/base/serializers.py:1442: in run_validation
    _validated_data = super(JSONAPISerializer, self).run_validation(data)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/serializers.py:426: in run_validation
    value = self.to_internal_value(data)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/serializers.py:483: in to_internal_value
    validated_value = field.run_validation(primitive_value)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/relations.py:153: in run_validation
    return super().run_validation(data)
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/rest_framework/fields.py:568: in run_validation
    value = self.to_internal_value(data)
api/draft_registrations/serializers.py:30: in to_internal_value
    node = self.context['view'].get_node(node_id=node_id) if node_id else None
E   AttributeError: 'DraftRegistrationDetail' object has no attribute 'get_node'
=============================== warnings summary ===============================
```

Reply:
Ah, thanks! So is this because `self.context['view']` return the wrong object `DraftRegistrationDetail`? Or it returns the correct object but DRF upgrade cause `DraftRegistrationDetail` to behave buggy?

Reply:
In fact, `DraftRegistrationDetail` should have access to the `.get_node()` either from its parent `DraftRegistrationSerializer` directly or from `DraftRegistrationLegacySerializer` via another parent `DraftRegistrationDetailLegacySerializer`. This looks like related to the MRO / DRF change.